pkg/k8s/watchers/crd.go

```diff
@@ -0,0 +1,299 @@
+// "k8s.io/client-go/tools/cache" package, with many alterations made to
+// efficiently retrieve Cilium CRDs. This is important because we don't want
+// each agent to fetch the full CRDs across the cluster, because they
+// potentially contain large validation schemas.
+//
+// This function also removes removes unnecessary calls from the upstream
+// version that set the namespace and the resource when performing `Get`.
+//
+//   - If the resource was set, the following error was observed:
+//     "customresourcedefinitions.apiextensions.k8s.io
+//     "customresourcedefinitions" not found".
+//   - If the namespace was set, the following error was observed:
+//     "an empty namespace may not be set when a resource name is provided".
+//
+// The namespace problem can be worked around by using NamespaceIfScoped, but
+// it's been omitted entirely here because it's equivalent in functionality.
+func newListWatchFromClient(c cache.Getter, fieldSelector fields.Selector) *cache.ListWatch {
+	optionsModifier := func(options *metav1.ListOptions) {
+		options.FieldSelector = fieldSelector.String()
+	}
+
+	listFunc := func(options metav1.ListOptions) (runtime.Object, error) {
+		// This lister  will retrieve the CRDs as a metav1.Table object. This
+		// is the same way that `kubectl get crds` returns them. In the
+		// metav1.Table object, it contains cells, which are
+		// metav1.PartialObjectMetadata objects containing the minimal
+		// representation of an objects in K8s (in this case a CRD). Once the
+		// metav1.Table is fetched, it is converted into a
+		// metav1.PartialObjectMetadataList which is what the controller
+		// (informer) expects the object type to be. If we returned the
+		// metav1.Table itself, the controller doesn't know how to handle that
+		// object type because it is not a list type.
+
+		optionsModifier(&options)
+
+		t := &metav1.Table{}
+		err := c.Get().
+			SetHeader("Accept", tableHeader).
+			VersionedParams(&options, metav1.ParameterCodec).
+			Do(context.TODO()).
+			Into(t)
+		if err != nil {
+			return nil, err
+		}
+		return tableToPomList(t), nil
+	}
+	watchFunc := func(options metav1.ListOptions) (watch.Interface, error) {
+		// This watcher will retrieve each CRD that the lister has listed as
+		// individual metav1.PartialObjectMetadata because it is requesting the
+		// apiserver to return objects as such via the "Accept" header.
+
+		optionsModifier(&options)
+
+		options.Watch = true
+		return c.Get().
+			SetHeader("Accept", partialObjHeader).
+			VersionedParams(&options, metav1.ParameterCodec).
+			Watch(context.TODO())
+	}
+	return &cache.ListWatch{ListFunc: listFunc, WatchFunc: watchFunc}
+}
+
+// tableToPomList converts a metav1.Table's cells into individual
+// metav1.PartialObjectMetadata objects, and returns them all placed in a
+// metav1.PartialObjectMetadataList.
+func tableToPomList(t *metav1.Table) *metav1.PartialObjectMetadataList {
```

Comment:
Small nit. This is a `metav1.Table` -> `metav1.POM`. We likely need another function that does `v1beta1.Table` to `v1beta1.POM`. Unless the v1.Table is available since k8s >= 1.12

Reply:
Addressed by putting a comment on why we may not need to do this.